builtin/logical/database/credentials.go

```diff
@@ -170,3 +175,188 @@ func (kg rsaKeyGenerator) configMap() (map[string]interface{}, error) {
+	}
+
+	if cg.CommonNameTemplate == "" {
+		return cg, fmt.Errorf("missing required common_name_template")
+	}
+
+	// Validate the common name template
+	t, err := template.NewTemplate(template.Template(cg.CommonNameTemplate))
+	if err != nil {
+		return cg, fmt.Errorf("failed to create template: %w", err)
+	}
+
+	_, err = t.Generate(dbplugin.UsernameMetadata{})
+	if err != nil {
+		return cg, fmt.Errorf("invalid common_name_template: %w", err)
+	}
+	cg.cnProducer = t
+
+	if cg.CACert == "" {
+		return cg, fmt.Errorf("missing required ca_cert")
+	}
+	if cg.CAPrivateKey == "" {
+		return cg, fmt.Errorf("missing required ca_private_key")
+	}
+	parsedBundle, err := certutil.ParsePEMBundle(strings.Join([]string{cg.CACert, cg.CAPrivateKey}, "\n"))
+	if err != nil {
+		return cg, err
+	}
+	if parsedBundle.PrivateKey == nil {
+		return cg, fmt.Errorf("private key not found in the PEM bundle")
+	}
+	if parsedBundle.PrivateKeyType == certutil.UnknownPrivateKey {
+		return cg, fmt.Errorf("unknown private key found in the PEM bundle")
+	}
+	if parsedBundle.Certificate == nil {
+		return cg, fmt.Errorf("certificate not found in the PEM bundle")
+	}
+	if !parsedBundle.Certificate.IsCA {
+		return cg, fmt.Errorf("the given certificate is not marked for CA use")
+	}
+
+	certBundle, err := parsedBundle.ToCertBundle()
+	if err != nil {
+		return cg, fmt.Errorf("error converting raw values into cert bundle: %w", err)
+	}
+
+	parsedCABundle, err := certBundle.ToParsedCertBundle()
+	if err != nil {
+		return cg, fmt.Errorf("failed to parse cert bundle: %w", err)
+	}
+	cg.parsedCABundle = parsedCABundle
+
+	return cg, nil
+}
+
+func (cg *ClientCertificateGenerator) generate(r io.Reader, expiration time.Time, userMeta dbplugin.UsernameMetadata) (*certutil.CertBundle, string, error) {
+	commonName, err := cg.cnProducer.Generate(userMeta)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Set defaults
+	keyBits := cg.KeyBits
+	signatureBits := cg.SignatureBits
+	switch cg.KeyType {
+	case "rsa":
+		if keyBits == 0 {
+			keyBits = 2048
+		}
+		if signatureBits == 0 {
+			signatureBits = 256
+		}
+	case "ec":
+		if keyBits == 0 {
+			keyBits = 256
+		}
+		if signatureBits == 0 {
+			signatureBits = keyBits
+		}
+	case "ed25519":
+		// key_bits ignored
+		if signatureBits == 0 {
+			signatureBits = 256
+		}
+	}
+
+	subject := pkix.Name{
+		CommonName: commonName,
+		// Additional subject DN options intentionally omitted for now
+	}
+
+	creation := &certutil.CreationBundle{
+		Params: &certutil.CreationParameters{
+			Subject:                       subject,
+			KeyType:                       cg.KeyType,
+			KeyBits:                       cg.KeyBits,
+			SignatureBits:                 cg.SignatureBits,
+			NotAfter:                      expiration,
+			KeyUsage:                      x509.KeyUsageDigitalSignature,
+			ExtKeyUsage:                   certutil.ClientAuthExtKeyUsage,
+			BasicConstraintsValidForNonCA: true,
+			NotBeforeDuration:             0,
```

Comment:
Might be worth a non-zero value here to back date it just slightly (`30s` is the default in PKI) just in case times are out of sync between Vault, the client, and MongoDB.

Reply:
Done in https://github.com/hashicorp/vault/pull/20425/commits/3e22a33ac4fe012634d2ba58003b1282b13baf48. Thanks!